### PR TITLE
Refactor to use IAsyncEnumerable

### PR DESCRIPTION
**Pull request by HowardvanRooijen**

### CODE REVIEW

File: Solutions/Ais.Net.Receiver/NmeaReceiver.cs [hunk 1]
```diff
@@ -2,99 +2,61 @@
...
 
+            int retryAttempt = 0;
 
-        public async IAsyncEnumerable<string> GetAsync()
-        {
-            using (var reader = new StreamReader(this.tcpClient.GetStream()))
+            while (this.tcpClient.Connected)
             {
-                while (true)
+                while (stream.DataAvailable)
```
Comment (idg10) [inline Solutions/Ais.Net.Receiver/NmeaReceiver.cs:45]: I was staring baffled at this code for a while trying to understand how it handles retries. Initially I was expecting a `catch` block and retry logic in the event of failures. When I didn't find such a thing I wondered if maybe "retry" meant something else here—perhaps with this source we read all the data it has, and then it stops, and then to get the further data we have to wait a bit and then ask it if it's got anything new?

I don't think it's that.

My current provisional conclusion is that the failure mode we see in practice is that the connection gets closed, and so we fall out of this loop.

Anyway, in conclusion, I'd really like a comment here explaining what failure looks like.

File: Solutions/Ais.Net.Receiver/ReceiverHost.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                TimeSpan.Parse(this.configuration[ConfigurationKeys.ConnectionRetryPeriodicity]),
+                int.Parse(this.configuration[ConfigurationKeys.ConnectionRetryAttempts]));
+
+            IObservable<string> messages = this.GetAsync().ToObservable();
+            await messages.Buffer(int.Parse(this.configuration[ConfigurationKeys.WriteBatchSize])).Select(this.storageClient.PersistAsync);
+        }
+
+        public async IAsyncEnumerable<string> GetAsync()
+        {
+            while (!this.receiver.Connected)
```
Comment (idg10) [inline Solutions/Ais.Net.Receiver/ReceiverHost.cs:41]: This baffles me. We read messages for as long as we are _not_ connected?

File: Solutions/Ais.Net.Receiver/ReceiverHost.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        public ReceiverHost(IConfiguration configuration, IStorageClient storageClient)
+        {
+            this.configuration = configuration;
+            this.storageClient = storageClient;
+        }
+
+        public async Task StartAsync()
+        {
+            this.receiver = new NmeaReceiver(
+                this.configuration[ConfigurationKeys.Host],
```
Comment (idg10) [inline Solutions/Ais.Net.Receiver/ReceiverHost.cs:30]: FWIW, I much prefer configuration binding for this sort of thing. If you defined a class with properties for each of these settings you'd have two benefits:

1. you wouldn't have to write all these calls to `Parse` yourself
2. you have a class whose structure tells you exactly what structure of config is required, along with the configuration keys, all in once place (so I don't have to look here to see the expected format, and then in `ConfigurationKeys` to find out the setting name; I'd just have a `public int Host { get; set; }` property that tells me both.

or does `TimeSpan` not work with config binding? Or is it that you particularly wanted these unixy names (all lowercase and dashes) and I'm not sure you can actually do that with config binding (not the dashes anyway).

File: Solutions/Ais.Net.Receiver/ReceiverHost.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                int.Parse(this.configuration[ConfigurationKeys.HostPort]),
+                TimeSpan.Parse(this.configuration[ConfigurationKeys.ConnectionRetryPeriodicity]),
+                int.Parse(this.configuration[ConfigurationKeys.ConnectionRetryAttempts]));
+
+            IObservable<string> messages = this.GetAsync().ToObservable();
```
Comment (idg10) [inline Solutions/Ais.Net.Receiver/ReceiverHost.cs:35]: So having gone to great trouble to replace Rx with `IAsyncEnumerable<T>` we finally turn it back into Rx again here. Why did we need to convert everything else to `IAsyncEnumerable<T>`?

File: Solutions/Ais.Net.Receiver/ReceiverHost.cs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+            await messages.Buffer(int.Parse(this.configuration[ConfigurationKeys.WriteBatchSize])).Select(this.storageClient.PersistAsync);
```
Comment (idg10) [inline Solutions/Ais.Net.Receiver/ReceiverHost.cs:36]: Eew! If I'm reading this right, you're using the `Select` operator for its side effects?

Even if we ignore how wrong this is from mathematical elegance point of view, there's a serious practical problem here: `PersistAsync` returns a `Task` which you are ignoring. This causes two problems:

1. failures will go unnoticed, because nothing ever inspects the task returned by `PersistAsync`
2. you might end up with multiple concurrent batch writes in progress simultaneously

Even in a world without errors, 2 looks like it could be a serious problem here. You're using append operations on a blob, right? In blob storage, we group things by hour, right? I'm assuming `WriteBatchSize` here is expected to produce a chunk that is smaller than 1 hour, right? So in normal operation, shortly after the top of an hour you'll create a new blob, and then you expect to append to that blob multiple times over the course of the hour, right?

But nothing here will ensure that these appends happen in order. If you happen to receive enough of a splurge of data from the source (or if blob storage has a temporary perf issue) it's possible that you will start attempting a new append before the previous batch's append has finished.

I was going to say that you probably want to use Rx's `ForEachAsync` operator here, except it, ironically, doesn't actually have support for `async`... (Well, it sort of does. It returns something you can `await` so you've got an `async`-friendly way of saying "I want to wait until you've processed everything here" but the callback continues to be an `Action` and it doesn't have any `Func<Task>` forms to enable an async handler. The reason for that is that `IObserver.OnNext` is also non-async. Possibly one of Rx's async dimensions could work here—I think they have something like `IAsyncObservable<T>`, which I'm guessing would have this feature.)

However, maybe this is `IAsyncEnumerable<T>`'s time to shine. Instaed of calling `ToObservable` above, just bring in the library that adds the full set of Rx/Ix operators to `IAsyncEnumerable`, and this inclues the `Buffer` that you're using here. You could then write an `await foreach` over that, enabling you to enforce the "one batch at a time" behaviour that I'm pretty sure you want here.

File: Solutions/Ais.Net.Receiver/IStorageClient.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+// </copyright>
+
+namespace Ais.Net.Receiver
+{
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+
+    public interface IStorageClient
+    {
+        Task PersistAsync(IEnumerable<string> messages);
```
[the remark this answers is not in the document]
Comment (idg10) [reply Solutions/Ais.Net.Receiver/IStorageClient.cs:12]: Actually, having read through the rest of the code, I now wonder if this should be `ICollection<string>`, to better reflect the fact that your only implementation of this that it expects the relevant data to be available immediately. The nature of this particular method is that it doesn't really want the source data to be lazy. (And you're feeding it with the output of the `Buffer` operator, which, IIRC, actually supplies you with an `IList<string>`, because by definition, `Buffer` only supplies you with fully-populated buffers.)

File: Solutions/Ais.Net.Receiver/StorageClient.cs [hunk 7]
```diff
@@ -15,33 +16,58 @@ namespace Endjin.Ais.Receiver
...
+        public async Task PersistAsync(IEnumerable<string> messages)
         {
-            await this.InitialiseConnectionAsync().ConfigureAwait(false);
+            await this.InitialiseContainerAsync().ConfigureAwait(false);
+            await using MemoryStream stream = new MemoryStream(Encoding.UTF8.GetBytes(messages.Aggregate(new StringBuilder(), (sb, a) => sb.AppendLine(string.Join(",", a)), sb => sb.ToString())));
+            await this.appendBlobClient.AppendBlockAsync(stream).ConfigureAwait(false);
+        }
 
-            using (var stream = new MemoryStream(Encoding.UTF8.GetBytes(messages.Aggregate(new StringBuilder(), (sb, a) => sb.AppendLine(String.Join(",", a)), sb => sb.ToString())))) 
+        private async Task InitialiseContainerAsync()
```
Comment (idg10) [inline Solutions/Ais.Net.Receiver/StorageClient.cs:37]: I think this is misnamed. You call it every time `PersistAsync` is called, so this doesn't really look like an initialization method.

File: Solutions/Ais.Net.Receiver/StorageClient.cs [hunk 8]
```diff
@@ -15,33 +16,58 @@ namespace Endjin.Ais.Receiver
...
+        {
+            try
+            {
+                this.blobContainerClient = new BlobContainerClient(
+                    this.configuration[ConfigurationKeys.ConnectionString],
+                    this.configuration[ConfigurationKeys.ContainerName]);
+                this.appendBlobClient = new AppendBlobClient(
+                    this.configuration[ConfigurationKeys.ConnectionString],
+                    this.configuration[ConfigurationKeys.ContainerName],
+                    $"raw/{DateTimeOffset.Now:yyyy}/{DateTimeOffset.Now:MM}/{DateTimeOffset.Now:dd}/{DateTimeOffset.Now:yyyyMMddTHH}.nm4");
```
Comment (idg10) [inline Solutions/Ais.Net.Receiver/StorageClient.cs:47]: Three problems here.

1. this partitions data by when we happened to write it out to blob storage, and now by when we received it (possibly not a big problem)
2. You should use `DateTimeOffset.UtcNow` because `Now` will reflect daylight saving, and since you're not writing the timezone out you'll encounter the "this hour happened twice" problem when the clocks go back.
3. `DateTimeOffset.Now` is a moving target, so it's possible that the filename won't match the folder structure

For 2, e.g. suppose it's 2021/02/28 23:59:59.99999 at the point where you start running this. So your expression here:

```cs
$"raw/{DateTimeOffset.Now:yyyy}/{DateTimeOffset.Now:MM}/{DateTimeOffset.Now:dd}/{DateTimeOffset.Now:yyyyMMddTHH}.nm4"
```

starts by reading the `DateTimeNow` and gets a year of 2021, then it reads it again and gets a month of 02, but then before it manages to read it for the third time, the clock rolls past midnight, so now you get a day of 01, because it's March now. So you end up with this sort of filename:

`raw/2021/02/01/2021030100.nm4`

For this sort of thing you always want to read `DateTime.UtcNow` once, store the value in a local, and then refer to that local in all the places where you want to pull out the component parts.

The bug you have here is particularly infuriating because you will _almost_ never encounter it in reality, making it incredibly baffling when you finally do. (And as my father learned from spending half his career in real-time systems, timing bugs will always hit you eventually. It's not good enough to make them very unlikely: that just makes them harder to repro.)

File: Solutions/Ais.Net.Receiver/StorageClient.cs [hunk 9]
```diff
@@ -15,33 +16,58 @@ namespace Endjin.Ais.Receiver
...
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine(e);
```
Comment (idg10) [inline Solutions/Ais.Net.Receiver/StorageClient.cs:51]: I'm guessing this was here because you noticed that you were never actually seeing the errors. But it was only necessary because of the abuse of the `Select` operator earlier on. If you change to an `await foreach` to write out the buffers, the exception should bubble out in to that code properly.

File: Solutions/Ais.Net.Receiver/NmeaReceiver.cs (hunk 1 above)
Comment (HowardvanRooijen) [reply Solutions/Ais.Net.Receiver/NmeaReceiver.cs:45]: When I last tried to run the code in 2019, we would occasionally get into a state where the remote AIS Sender process would become unresponsive, but the host was still resolvable, and DataAvailable was the only mechanism for detecting that.

**Comment by HowardvanRooijen** [comment]
Code integrated via another branch
